docs/design/variadics.md

```diff
@@ -0,0 +1,947 @@
+implemented at monomorphization time, so all types are known constants, and all
+all arities are known.
+
+A tuple pattern can contain no more than one subpattern of the form "`...`
+_operand_". When such a subpattern is present, the N elements of the pattern
+before the `...` expansion are matched with the first N elements of the
+scrutinee, and the M elements of the pattern after the `...` expansion are
+matched with the last M elements of the scrutinee. If the scrutinee does not
+have at least N + M elements, the pattern does not match.
+
+The remaining elements of the scrutinee are iteratively matched against
+_operand_, in order. In each iteration, `$I` is equal to the index of the
+scrutinee element being matched, minus N.
+
+On the Nth iteration, a binding pattern binds the Nth element of the named pack
+to the Nth scrutinee value.
+
+## Typechecking
+
+### Tuples, packs, segments, and shapes
+
+In order to discuss the underlying type system for variadics, we will need to
+introduce some pseudo-syntax to represent values and expressions that occur in
+the type system, but cannot be expressed directly in user code. We will use
+non-ASCII glyphs such as `⟪⟫‖⟬⟭` for that pseudo-syntax, to distinguish it from
+valid Carbon syntax.
+
+In the context of variadics, we will say that a tuple literal consists of a
+comma-separated sequence of _segments_, and reserve the term "elements" for the
+components of a tuple literal after pack expansion. For example, the expression
+`(... each foo)` may evaluate to a tuple value with any number of elements, but
+the expression itself has exactly one segment.
+
+The type of a tuple literal is another tuple literal, which is computed
+segment-wise. For example, suppose we are trying to find the type of `z` in this
+code:
+
+```carbon
+fn F[... each T:! type]((... each x: Optional(each T)), (... each y: i32)) {
+  let z: auto = (0 as f32, ... each x, ... each y);
+}
+```
+
+We proceed by finding the type of each segment. The type of `0 as f32` is `f32`,
+by the usual non-variadic typing rules. The type of `... each x` is
+`... Optional(each T)`, because `Optional(each T)` is the declared type of
+`each x`, and the type of a pack expansion is a pack expansion of the type of
+its body.
+
+The type of `... each y` is more complicated. Conceptually, it consists of some
+number of repetitions of `i32`. We don't know exactly how many repetitions,
+because it's implicitly specified by the caller: it's the arity of the second
+argument tuple. Effectively, that arity acts as a hidden deduced parameter of
+`F`.
+
+So to represent this type, we need two new pseudo-syntaxes:
+
+-   `‖each X‖` refers to the deduced arity of the pack expansion that contains
+    the declaration of `each X`.
+-   `⟪E; N⟫` evaluates to `N` repetitions of `E`. This is called a _arity
+    coercion_, because it coerces the expression `E` to have arity `N`. `E` must
+    not contain any pack expansions, each-names, or pack literals (see below).
```

Comment:
Let me try this:

> -   `‌‖‌each X‌‖‌` refers to the deduced arity of the pack expansion that contains
>     the declaration of `each X`.
> -   `‌⟪‌E; N‌⟫‌` evaluates to `N` repetitions of `E`. This is called a _arity
>     coercion_, because it coerces the expression `E` to have arity `N`. `E` must
>     not contain any pack expansions, each-names, or pack literals (see below).
> 
> Combining the two, the type of `... each y` is `... ‌⟪‌i32; ‌‖‌each y‌‖‌‌⟫‌`. Thus, the
> type of `z` is `(f32, ... Optional(each T), ... ‌⟪‌i32; ‌‖‌each y‌‖‌‌⟫‌)`.
> 
> Now, consider a modified version of that example:
> 
> ```carbon
> fn F[... each T:! type]((... each x: Optional(each T)), (... each y: i32)) {
>   let (.. each z: auto) = (0 as f32, ... each x, ... each y);
> }
> ```
> 
> `each z` is a pack, but it has the same elements as the tuple `z` in our earlier
> example, so we represent its type in the same way, as a sequence of segments:
> `‌⟬‌f32, Optional(each T), ‌⟪‌i32; ‌‖‌each y‌‖‌‌⟫‌‌⟭‌`. The `‌⟬‌‌⟭‌` delimiters make this a
> _pack literal_ rather than a tuple literal. Notice one subtle difference: the
> segments of a pack literal do not contain `...`. In effect, every segment of a
> pack literal acts as a separate loop body. As with the tuple literal syntax, the
> pack literal pseudo-syntax can also be used in patterns.
> 
> The _shape_ of a pack literal is a tuple of the arities of its segments, so the
> shape of `‌⟬‌f32, Optional(each T), ‌⟪‌i32; ‌‖‌each y‌‖‌‌⟫‌‌⟭‌` is
> `(1, ‌‖‌each T‌‖‌, ‌‖‌each y‌‖‌)`. Other expressions and patterns also have shapes. In
> particular, the shape of an arity coercion `‌⟪‌E; A‌⟫‌` is `(A)`, the shape of
> `each X` is `‌‖‌each X‌‖‌`, and the shape of an expression that does not contain

That seems to be working for me. What I did was to insert a U+200C (zero width non-joiner) on both sides of each of the fancy characters. Unfortunately that seems to give the worse rendering of the double-chevron. I wonder if using guillemets (U+AB, U+BB: « ») instead of the mathematical double-angle brackets would work better:


> `‌⟬‌f32, Optional(each T), ‌«i32; ‌‖‌each y‌‖‌‌»⟭‌`. The `‌⟬‌‌⟭‌` delimiters make this a


That seems OK. Maybe there's a better character to use here, though.